omegaconf/dictconfig.py

```diff
@@ -186,12 +188,19 @@ def is_typed(c: Any) -> bool:
         if is_dict(value_type) and is_dict(target_type):
             return
 
+        def is_not_valid_type(value_type: Any, target_type: Any) -> bool:
+            return (
+                type_is_primitive_list(value_type)
+                or is_primitive_type(value_type)
+                or not issubclass(value_type, target_type)
+            )
+
```

Comment:
I don't understand the logic here.
why is a list not valid? why is a primitive not valid?

Reply:
Checking if issubclass(value_type, target_type) fails every time primitive types are passed.

Reply:
Can you explain the additional check for is_primitive_list?
won't issubclass catch that? what case is this handling?

Reply:
```
@dataclass
class ListClass:
    list: List[int] = field(default_factory=lambda: []) 
    tuple: Tuple[int, int] = field(default_factory=lambda: (1, 1)) 
    dict: Dict[str, Any] = field(default_factory=lambda: {"foo": "bar"})


cfg = OmegaConf.structured(ListClass)
cfg.dict =  1 # fails
cfg.dict = [1, True] # fails
```
Both assignments end up with 
```
Traceback (most recent call last):
  File "/Users/peristocles/contributions/omegaconf/omegaconf/dictconfig.py", line 302, in __setattr__
    self.__set_impl(key, value)
  File "/Users/peristocles/contributions/omegaconf/omegaconf/dictconfig.py", line 286, in __set_impl
    self._set_item_impl(key, value)
  File "/Users/peristocles/contributions/omegaconf/omegaconf/basecontainer.py", line 376, in _set_item_impl
    self._validate_set(key, value)
  File "/Users/peristocles/contributions/omegaconf/omegaconf/dictconfig.py", line 137, in _validate_set
    self._validate_set_merge_impl(key, value, is_assign=True)
  File "/Users/peristocles/contributions/omegaconf/omegaconf/dictconfig.py", line 203, in _validate_set_merge_impl
    and not issubclass(value_type, target_type)
  File "/Users/peristocles/anaconda3/envs/omegaconf/lib/python3.6/typing.py", line 1148, in __subclasscheck__
    raise TypeError("Parameterized generics cannot be used with class "
TypeError: Parameterized generics cannot be used with class or instance checks
```

Reply:
I see, but why are you checking specifically list and not dict?

I just ran into something related in Hydra:
```
# Allows a list to be assigned to an int:
python examples/tutorials/structured_configs/1_minimal/my_app.py 'port=[1,2]'
Host: localhost, port: [1, 2]

# But correctly rejects a float:
omry@Coronadev:~/dev/hydra$ python examples/tutorials/structured_configs/1_minimal/my_app.py port=1.0
Error merging override port=1.0
Value '1.0' could not be converted to Integer
        full_key: port
        reference_type=Optional[MySQLConfig]
        object_type=MySQLConfig
```

Can you confirm that this works correctly after this fix?

Reply:
> I see, but why are you checking specifically list and not dict?

Dict is checked before, lists raise the same error as primitive types.

> I just ran into something related in Hydra:
> 
> ```
> # Allows a list to be assigned to an int:
> python examples/tutorials/structured_configs/1_minimal/my_app.py 'port=[1,2]'
> Host: localhost, port: [1, 2]
> 
> # But correctly rejects a float:
> omry@Coronadev:~/dev/hydra$ python examples/tutorials/structured_configs/1_minimal/my_app.py port=1.0
> Error merging override port=1.0
> Value '1.0' could not be converted to Integer
>         full_key: port
>         reference_type=Optional[MySQLConfig]
>         object_type=MySQLConfig
> ```
> 
> Can you confirm that this works correctly after this fix?

It isn't working I'll check what is happening. This fix only targeted containers.

Reply:
> Dict is checked before, lists raise the same error as primitive types.
Okay, maybe check list before as well? the code now is pretty hard to parse.

> It isn't working I'll check what is happening. This fix only targeted containers.
Probably better to file a different bug.

Reply:
> Okay, maybe check list before as well? the code now is pretty hard to parse.

I've been looking to simplify it and, when issubclass fails is because it doesn't support subscripted generics like `List[Any]` or  `Dict[str, int]`

There is a post about how to  [check if a type is a subscripted generic](https://stackoverflow.com/questions/49171189/whats-the-correct-way-to-check-if-an-object-is-a-typing-generic) which results in adding this logic removing previous changes I did but, I would mean adding that chuck of code in _utils.

```
if is_generic(target_type) and not is_base_generic(target_type):
            return
```

Reply:
I've tried it and it solves the problem.

Reply:
There are already functions in utils to check if something is a Dict or a List (is_dict_annotation, is_list_annotation), as well as functions to check if something is a primitive dict or list.

Reply:
I know but they don't check what is needed. For example for `List[Any]` and `List` in is_list_annotation would return true, but I need only to check if is like `List[Any], 
Dict[str, int]`...

Reply:
Do we ever use List without  the type (List[T] )?

Reply:
No. But I guess it could happen